antenna/app.py

```diff
@@ -3,223 +3,211 @@
+            # the payload size by decompressing it. We save the original value
+            # in case we need to debug something later on.
+            req.env['ORIG_CONTENT_LENGTH'] = content_length
+            req.env['CONTENT_LENGTH'] = len(data)
+
+            data = io.BytesIO(data)
         else:
             data = req.stream
 
-        env = req.env.copy()
-        fs = cgi.FieldStorage(fp=data, environ=env, keep_blank_values=1)
-        form = Storage(
-            [(k, self._process_fieldstorage(fs[k])) for k in fs.keys()]
-        )
-        # FIXME: In the original collector, this returned request querystring
-        # data as well as request body data.
-        return form
-
-    @staticmethod
-    def _no_x00_character(value):
-        """Remove x00 characters
-
-        Note: We remove null characters because they are a hassle to deal with
-        during reporting and cause problems when sending to Postgres.
-
-        :arg value: a basestring with null characters in it
-
-        :returns: same type basestring with null characters removed
-
-        """
-        if isinstance(value, unicode) and u'\u0000' in value:
-            return u''.join(c for c in value if c != u'\u0000')
-        if isinstance(value, str) and '\x00' in value:
-            return ''.join(c for c in value if c != '\x00')
-        return value
-
-    def _get_raw_crash_from_form(self, req):
-        """Retrieves crash/dump data and fixes it
-
-        :arg req: the WSGI request
+        # Convert to FieldStorage and then to dict.
+        fs = cgi.FieldStorage(fp=data, environ=req.env, keep_blank_values=1)
+        payload = self.process_fieldstorage(fs)
 
-        :returns: (raw_crash, dumps)
+        # NOTE(willkg): In the original collector, this returned request
+        # querystring data as well as request body data, but we're not doing
+        # that because the query string just duplicates data in the payload.
 
-        """
-        dumps = MemoryDumpsMapping()
         raw_crash = {}
-        checksums = {}
-
-        for name, value in self._form_as_mapping(req).iteritems():
-            name = self._no_x00_character(name)
-            if isinstance(value, basestring):
-                if name != "dump_checksums":
-                    raw_crash[name] = self._no_x00_character(value)
-            elif hasattr(value, 'file') and hasattr(value, 'value'):
-                dumps[name] = value.value
-                checksums[name] = self.checksum_method(value.value).hexdigest()
-            elif isinstance(value, int):
-                raw_crash[name] = value
+        dumps = {}
+
+        # FIXME(willkg): I think this has extra stanzas it doesn't need. Pretty
+        # sure payload items are either strings or bytes and that's it. I threw
+        # some asserts in there that will fail immediately, so if we ever bump
+        # into it, we'll know. This should reveal itself in testing. If it
+        # never pops up, we can nix the sections.
```

Comment:
I spent some quality time looking at existing raw crashes, the Python FieldStorage code and sending stuff in to see what happens and fixed this up.